File: apps/web/lib/server/engine/benchmark.ts
/**
 * Benchmark metrics (spec §8). Pure functions so the gate is deterministic and
 * unit-testable independently of the data. The gate itself (Task 9) runs the
 * full blend path against committed fixtures and applies thresholds.
 */
export interface BenchmarkOutcome {
  id: string;
  product_category: string;
  outcome: { label: "hit" | "moderate" | "flop"; known_failure_modes?: string[] };
}

export function outcomeRank(label: "hit" | "moderate" | "flop"): number {
  return label === "hit" ? 2 : label === "moderate" ? 1 : 0;
}

function rankdata(xs: number[]): number[] {
  const idx = xs.map((v, i) => [v, i] as const).sort((a, b) => a[0] - b[0]);
  const ranks = new Array<number>(xs.length);
  let i = 0;
  while (i < idx.length) {
    let j = i;
    while (j + 1 < idx.length && idx[j + 1][0] === idx[i][0]) j++;
    const avg = (i + j) / 2 + 1; // average rank for ties, 1-based
    for (let k = i; k <= j; k++) ranks[idx[k][1]] = avg;
    i = j + 1;
  }
  return ranks;
}

export function spearman(a: number[], b: number[]): number {
  if (a.length !== b.length || a.length < 2) return 0;
  const ra = rankdata(a);
  const rb = rankdata(b);
  const n = a.length;
  const mean = (n + 1) / 2;
  let num = 0, da = 0, db = 0;
  for (let i = 0; i < n; i++) {
    num += (ra[i] - mean) * (rb[i] - mean);
    da += (ra[i] - mean) ** 2;
    db += (rb[i] - mean) ** 2;
  }
  return da === 0 || db === 0 ? 0 : num / Math.sqrt(da * db);
}

export function failureModeHitRate(results: { known: string[]; topBlockers: string[] }[]): number {
  const withModes = results.filter((r) => r.known.length > 0);
  if (withModes.length === 0) return 1;
  const hits = withModes.filter((r) => r.known.some((m) => r.topBlockers.includes(m))).length;
  return hits / withModes.length;
}

export function withinCategoryInversions(rows: { category: string; score: number; rank: number }[]): number {
  let inversions = 0;
  const byCat = new Map<string, { score: number; rank: number }[]>();
  for (const r of rows) {
    if (!byCat.has(r.category)) byCat.set(r.category, []);
    byCat.get(r.category)!.push(r);
  }
  for (const group of byCat.values()) {
    for (let i = 0; i < group.length; i++) {
      for (let j = 0; j < group.length; j++) {
        if (group[i].rank > group[j].rank && group[i].score < group[j].score) inversions++;
      }
    }
  }
  return inversions;
}

File: apps/api/app/services/benchmark.py
"""Benchmark metrics (spec §8) — Python mirror of
apps/web/lib/server/engine/benchmark.ts.

Pure functions so the backtest gate (apps/api/tests/test_benchmark_backtest.py)
is deterministic and unit-testable independently of the data. The gate itself
runs the full blend path against committed recorded fixtures and applies
thresholds measured on this engine's output (see the gate module docstring).
"""

from __future__ import annotations


def outcome_rank(label: str) -> int:
    return 2 if label == "hit" else 1 if label == "moderate" else 0


def _rankdata(xs: list[float]) -> list[float]:
    """1-based ranks, averaging ties (SciPy's `rankdata(method="average")`)."""
    order = sorted(range(len(xs)), key=lambda i: xs[i])
    ranks = [0.0] * len(xs)
    i = 0
    while i < len(order):
        j = i
        while j + 1 < len(order) and xs[order[j + 1]] == xs[order[i]]:
            j += 1
        avg_rank = (i + j) / 2 + 1  # average rank for ties, 1-based
        for k in range(i, j + 1):
            ranks[order[k]] = avg_rank
        i = j + 1
    return ranks


def spearman(a: list[float], b: list[float]) -> float:
    if len(a) != len(b) or len(a) < 2:
        return 0.0
    ra = _rankdata(a)
    rb = _rankdata(b)
    n = len(a)
    mean = (n + 1) / 2
    num = 0.0
    da = 0.0
    db = 0.0
    for i in range(n):
        num += (ra[i] - mean) * (rb[i] - mean)
        da += (ra[i] - mean) ** 2
        db += (rb[i] - mean) ** 2
    return 0.0 if da == 0 or db == 0 else num / (da * db) ** 0.5


def failure_mode_hit_rate(results: list[dict]) -> float:
    """`results`: list of {"known": list[str], "topBlockers": list[str]}."""
    with_modes = [r for r in results if r["known"]]
    if not with_modes:
        return 1.0
    hits = sum(1 for r in with_modes if any(m in r["topBlockers"] for m in r["known"]))
    return hits / len(with_modes)


def within_category_inversions(rows: list[dict]) -> int:
    """`rows`: list of {"category": str, "score": float, "rank": int}."""
    inversions = 0
    by_cat: dict[str, list[dict]] = {}
    for r in rows:
        by_cat.setdefault(r["category"], []).append(r)
    for group in by_cat.values():
        for gi in group:
            for gj in group:
                if gi["rank"] > gj["rank"] and gi["score"] < gj["score"]:
                    inversions += 1
    return inversions
